MediaRecorder.bs

```diff
@@ -286,11 +276,30 @@ interface MediaRecorder : EventTarget {
     <li>Set |recorder|'s {{state}} to {{recording}}, and run the following steps
     in parallel:
     <ol>
-      <li>Start recording all tracks in |tracks| using the {{MediaRecorder}}'s
-      current configuration, and gather the data into a {{Blob}} <var>blob</var>
-      and queue a task, using the DOM manipulation task source, to
-      <a>fire an event</a> named <a>start</a> at <var>recorder</var>, then return
-      <code>undefined</code>.</li>
+      <li>As soon as data is available on all tracks in |tracks|, the User Agent
```

Comment:
This is weird because `As soon as data is available...` is not a step you can run. It's not something you introduced, it's just this spec's legacy.

Perhaps we could try to improve this at the same time, maybe by making it the below?
```
Set |recorder|'s {{state}} to {{recording}}, and as soon as data is available
on all tracks in |tracks|, run the following steps in parallel:
- Let |extendedMimeType| be the value of |recorder|'s
  [=[[ConstrainedMimeType]]=] slot.
- Modify |extendedMimeType| by adding media type, subtype and codecs
  parameter reflecting the configuration used by the MediaRecorder to record
  all tracks in |tracks|, if not already present. This MAY include the
  profiles parameter [[RFC6381]] or further codec-specific parameters.
- Set |recorder|'s {{mimeType}} attribute to |extendedMimeType|.
- Start gathering data into a {{Blob}} <var>blob</var>.
- <a>Fire an event</a> named <a>start</a> at |recorder|.
```

Followed by your note.

An issue I can think of however is: if one track doesn't have any data for a long time -- what happens to the data on other tracks? As this stands it will be discarded, but that would regress current MediaRecorder behavior which records it.

Could we solve that? Some sort of track-by-track approach perhaps?

@jan-ivar might have thoughts too, he's much more seasoned in spec writing than myself.

Reply:
> An issue I can think of however is: if one track doesn't have any data for a long time -- what happens to the data on other tracks? As this stands it will be discarded, but that would regress current MediaRecorder behavior which records it.
> 
> Could we solve that? Some sort of track-by-track approach perhaps?

This does change the spec to say not to record until you have data, which could cause a regression, however:
- Based on a brief chat with someone, I don't _think_ that current implementations actually start recording prior to having data - though note that I have not verified if this is true.
- I seriously doubt that applications are relying on recording starting even though their tracks are muted at the time of starting. What's the use case of this?

I prefer not making the API more complicated by having track-by-track basis of starting recording. Just like it is best to move updating mimeType to "onstart", I think it's best to move the start of recording to when we know how we want to record, even if that is a change in behavior.

Reply:
> > An issue I can think of however is: if one track doesn't have any data for a long time -- what happens to the data on other tracks? As this stands it will be discarded, but that would regress current MediaRecorder behavior which records it.
> > Could we solve that? Some sort of track-by-track approach perhaps?
> 
> This does change the spec to say not to record until you have data, which could cause a regression, however:
> * Based on a brief chat with someone, I don't _think_ that current implementations actually start recording prior to having data - though note that I have not verified if this is true.

Firefox will start as soon as all tracks are available. And with current WIP patches I have (making us compliant with the current spec), this means it will start synchronously in `start()`. I.e., it doesn't wait for data on all tracks -- any track that doesn't have data initially will get padding or a delayed start time as needed. It does wait for output from the muxer (i.e., it waits for data to get written to the blob) before firing "start".

> * I seriously doubt that applications are relying on recording starting even though their tracks are muted at the time of starting. What's the use case of this?

One use case it could allow in the future would be to let `Stop();Start();` if they happen in the same task, stop the current recording and start a new one in a seamless way. The full blobs produced by two such recordings could be played seamlessly with MSE. It would be one way to handle track-set changes on the fly.

With your proposed changes we lock ourselves out of that possibility.

> I prefer not making the API more complicated by having track-by-track basis of starting recording. Just like it is best to move updating mimeType to "onstart", I think it's best to move the start of recording to when we know how we want to record, even if that is a change in behavior.

Reply:
> One use case it could allow in the future would be to let Stop();Start(); if they happen in the same task, stop the current recording and start a new one in a seamless way. The full blobs produced by two such recordings could be played seamlessly with MSE. It would be one way to handle track-set changes on the fly.
> 
> With your proposed changes we lock ourselves out of that possibility.

I'm happy to lock ourselves out of that possibility. I don't think we should make an operation like "replace track" be implemented by a series of method calls relying on being invoked in the same task, it seems a bit implicit, affecting implementation strategies and possibly limiting or adding complexity, and API ergonomics-wise, a step like "stop" in the middle not being descriptive of what we're trying to achieve. Besides, if we achieve "replace track" by chaining two existing methods with existing uses, how do you feature detect for "replace track"?

Anyway that's my personal opinion. But at TPAC Fukuoka 2019 the W3C working group [consensus was reached](https://www.w3.org/2019/09/18-webrtc-minutes.html#item09) to add a replaceTrack() method (slides [76](https://docs.google.com/presentation/d/1Xprz0ElNTUWRDdX6kyp11cZ4ItTvUFUmzoXsKZdXxuU/edit#slide=id.g62a86e6d8c_3_12)-[77](https://docs.google.com/presentation/d/1Xprz0ElNTUWRDdX6kyp11cZ4ItTvUFUmzoXsKZdXxuU/edit#slide=id.g634b2fa483_43_0)) to handle that use case, so I think we're fine to do what we want here.

Reply:
I think the [resolution](https://www.w3.org/2019/09/18-webrtc-minutes.html#resolution09) was to *"Start with proposal B, Jan-Ivar to create a PR, promise based."* - I read that as there being interest in pursuing `replaceTrack`. We don't have a final PR to have consensus over.

Regardless, I think they solve different problems:
 1. `replaceTrack` seamlessly merges two sources into one container.
 2. `rec.stop(); rec.start()` splits one source into two (independently playable) containers.

Reply:
> > One use case it could allow in the future would be to let Stop();Start(); if they happen in the same task, stop the current recording and start a new one in a seamless way. The full blobs produced by two such recordings could be played seamlessly with MSE. It would be one way to handle track-set changes on the fly.
> > With your proposed changes we lock ourselves out of that possibility.
> 
> I'm happy to lock ourselves out of that possibility. I don't think we should make an operation like "replace track" be implemented by a series of method calls relying on being invoked in the same task

It wouldn't be "replace track". It would most likely be "add track" since that's the only thing that cannot be fulfilled by the current spec or the `replaceTrack` proposal.

> , it seems a bit implicit, affecting implementation strategies and possibly limiting or adding complexity, and API ergonomics-wise, a step like "stop" in the middle not being descriptive of what we're trying to achieve. Besides, if we achieve "replace track" by chaining two existing methods with existing uses, how do you feature detect for "replace track"?
> 
> Anyway that's my personal opinion. But at TPAC Fukuoka 2019 the W3C working group [consensus was reached](https://www.w3.org/2019/09/18-webrtc-minutes.html#item09) to add a replaceTrack() method (slides [76](https://docs.google.com/presentation/d/1Xprz0ElNTUWRDdX6kyp11cZ4ItTvUFUmzoXsKZdXxuU/edit#slide=id.g62a86e6d8c_3_12)-[77](https://docs.google.com/presentation/d/1Xprz0ElNTUWRDdX6kyp11cZ4ItTvUFUmzoXsKZdXxuU/edit#slide=id.g634b2fa483_43_0)) to handle that use case explicitly, so I think we're free to do what we want here.

Ok. If we can't solve this I'd rather see us skip the mimetype deferral and only add a note. The application can ensure the track has data with the "mute" event like you [mentioned before](https://github.com/w3c/mediacapture-record/issues/139#issuecomment-535480592), should it want the UA to be able to take the received-track's codec into account. It could of course also take a guess and hope that they match.